OpenChangePasswordDialog new rules #328

Drakonian opened the pull request:
<!-- Thank you for submitting a Pull Request. If you're new to contributing to BCApps please read our pull request guideline below
* https://github.com/microsoft/BCApps/Contributing.md
-->
#### Summary <!-- Provide a general summary of your changes -->

If OldPassword is passed, we check that the new password should match the OldPassword
Also in this case we check that the old password matches the entered old password
New test

#### Work Item(s) <!-- Add the issue number here after the #. The issue needs to be open and approved. Submitting PRs with no linked issues or unapproved issues is highly discouraged. -->
Fixes #328 





Fixes [AB#506715](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/506715)

grobyns commented on src/System Application/App/Password/src/PasswordDialog.Page.al:
```diff
@@ -27,6 +27,12 @@ page 9810 "Password Dialog"
                 Visible = ShowOldPassword;
+                trigger OnValidate()
+                begin
+                    if OldPasswordToCompare <> '' then
```
All logic should be moved the impl codeunit

grobyns commented on src/System Application/App/Password/src/PasswordDialog.Page.al:
```diff
@@ -85,12 +95,16 @@ page 9810 "Password Dialog"
         [NonDebuggable]
         OldPasswordValue: Text;
+        [NonDebuggable]
+        OldPasswordToCompare: Text;
```
why are these not SecretText?

Drakonian replied:
@grobyns
Actually the reason is that on the page the user cannot interact directly with SecretText, variables with type text are used for input. Therefore, it was more convenient to compare text with text at once. 

For the same reason, I decided to place the logic of checks directly on the page.
Anyway, I'll fix what you asked for )

Drakonian commented:
PR is updated

Drakonian commented on src/System Application/App/Password/src/PasswordDialogImpl.Codeunit.al:
```diff
@@ -99,15 +99,15 @@ codeunit 9811 "Password Dialog Impl."
     procedure ValidateOldPasswordMatch(OldPasswordToCompare: SecretText; OldPasswordEntered: SecretText)
     begin
-        if OldPasswordToCompare.Unwrap() <> '' then
+        if not OldPasswordToCompare.IsEmpty() then
```
It is better to use the built-in IsEmpty() function to check for emptiness

JesperSchulz commented:
Build is failing. Please update PR and let's do another run 😊

Drakonian commented:
Updated

JesperSchulz commented on src/System Application/App/Password/src/PasswordDialogImpl.Codeunit.al:
```diff
@@ -92,5 +95,21 @@ codeunit 9811 "Password Dialog Impl."
+    [NonDebuggable]
+    procedure ValidateNewPasswordUniqueness(OldPasswordToCompare: SecretText; NewPassword: SecretText)
+    begin
+        if not OldPasswordToCompare.IsEmpty() then
```
You could reduce nesting here by by doing an "if isempty then exit". Just a suggestion.

Drakonian replied:
I agree. I often write like that myself, don't know why I haven't done the same here :)

darjoo commented on src/System Application/App/Password/src/PasswordDialog.Page.al:
```diff
@@ -27,6 +27,10 @@ page 9810 "Password Dialog"
                 Visible = ShowOldPassword;
+                trigger OnValidate()
+                begin
+                    PasswordDialogImpl.ValidateOldPasswordMatch(OldPasswordToCompare, OldPasswordValue);
```
Shouldn't we also do the same check `if RequiresPasswordValidation then...` here?

Drakonian replied:
As far as I understood from the discussion of the problem it was decided that this rule should always be active. Besides, `RequiresPasswordValidation` is now used only to check the strength of the password and the user has the option not to enable it. 

Based on all this information, it seems to me that I don't need to be bound to `RequiresPasswordValidation`

JesperSchulz commented:
I'll do some functional testing of those changes early next week 🙂 If all goes well, we merge!

JesperSchulz commented on src/System Application/Test/Password/src/PasswordDialogTest.Codeunit.al:
```diff
@@ -243,14 +269,29 @@ codeunit 135033 "Password Dialog Test"
     end;
 
+    [ModalPageHandler]
+    procedure ChangePasswordDialogWitldOldPasswordModalPageHandler(var PasswordDialog: TestPage "Password Dialog");
```
There seems to be some typo in the procedure name?
...WitldOld...

JesperSchulz commented on src/System Application/Test/Password/src/PasswordDialogTest.Codeunit.al:
```diff
@@ -221,6 +224,29 @@ codeunit 135033 "Password Dialog Test"
+        Password: SecretText;
+        OldPassword: SecretText;
+    begin
+        // [SCENARIO] Open Password dialog in change password mode. Old password
```
Old password?

Drakonian replied:
Typo removed, thanks

JesperSchulz reviewed (dismissed):
Looks good to me. Manual tests conducted successfully.

Drakonian commented:
Something's broken, the build's been going on for 3 hours.
![image](https://github.com/microsoft/BCApps/assets/16802407/c9215212-d965-4e9c-8697-28d7f2ba640e)

JesperSchulz commented:
> Something's broken, the build's been going on for 3 hours. ![image](https://private-user-images.githubusercontent.com/16802407/322872934-c9215212-d965-4e9c-8697-28d7f2ba640e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyODAzNjksIm5iZiI6MTcxMzI4MDA2OSwicGF0aCI6Ii8xNjgwMjQwNy8zMjI4NzI5MzQtYzkyMTUyMTItZDk2NS00ZTljLTg2OTctMjhkN2YyYmE2NDBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE2VDE1MDc0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxNTk4YmNjZDg2MzY4ZDEzMWI3OWI0YWVkZGEwNDk4OTVjMTJkODMzODQwMjQ4YTA3ZGIxYzQ3MjdiMjVlNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.B2ZQRJlge-Aj-6KzesXSf-jv49N7U58_k2wvCDhxbeU)

@mazhelez / @aholstrup1, you're on!

AndreasMoth commented on src/System Application/App/Password/src/PasswordDialogImpl.Codeunit.al:
```diff
@@ -73,6 +75,7 @@ codeunit 9811 "Password Dialog Impl."
         PasswordDialog: Page "Password Dialog";
     begin
         PasswordDialog.EnableChangePassword();
+        PasswordDialog.SetOldPasswordToCompareSecretValue(OldPassword);
```
The OldPassword parameter is a bit confusing now.. So as an input it's the old password that should be compared to, as an output it's the old password the user typed. Now these are not necessarily equal, if the OldPassword input was empty, then the OldPassword output is not necessarily empty. Of course if OldPassword input has a value, then OldPassword output are equal.

Drakonian replied:
Exactly, they can have differences and it's all because the original function had a very strange design where the OldPassword was just what the user entered in the dialog box, all the checks were kind of out of brackets. This in turn created confusion about how to use the function at all. 

https://github.com/microsoft/BCApps/issues/328#issuecomment-1983618268

To support backwards compatibility, when passing a blank to OldPassword (and this function was very often used this way) it will work as it did before the change. But only when OldPassword is filled in will the function actually perform a comparison. In addition, checking that the old password does not match the new one makes sense only when the old password has been sent. Otherwise there is nothing to check.

So the logic is as follows
Input OldPassword = '' works as before, we just return the entered password, no checks.
Input OldPassword <> '' if we do pass OldPassword it means that we assume a specific OldPassword and if it doesn't match the password the user fills in, an error is inevitable and necessary. Only in this case there is any chance to compare if OldPassword doesn't match NewPassword.

I'd say more confusing was how the function worked previously

AndreasMoth commented on src/System Application/App/Password/src/PasswordDialogManagement.Codeunit.al:
```diff
@@ -116,7 +116,7 @@ codeunit 9810 "Password Dialog Management"
     /// otherwise an empty text are returned.
     /// </summary>
-    /// <param name="OldPassword">Out parameter, the old password user typed on the dialog.</param>
+    /// <param name="OldPassword">Out parameter, the old password to compare with the password user typed on the dialog.</param>
```
We should specify OldPassword is also an input parameter. From this I expect the function itself determines the old password, compares and gives me the old password in case I want to do some kind of comparison.

Drakonian replied:
I agree now this is input and output parameter

Drakonian replied:
It probably makes no sense to return OldPassword at all, since we introduce checks and standards at the PasswordDialogManagment level.

Perhaps the best way would be to mark the old method as obsolete in future releases and create a new one (but would have to give a new name, number of parameters the same).

@AndreasMoth what do you think?

AndreasMoth replied:
Yes, that makes sense, to at least remove "var" from the OldPassword (through obsoletion, remember CLEAN tags).
You can also argue if we should even return the password, why does the caller need to get that password? But for now, just remove the "var" from OldPassword. We could also rename OldPassword to CurrentPassword while Password should be NewPassword (after all, you give the current password and get out the new password - renaming is not super important though).

Drakonian commented:
@AndreasMoth @JesperSchulz 

I've updated the code and updated the variable names.

I would like to point out that this strange design when OldPassword was returned was only needed for OnPrem report 9810 “Change Password”. 
The point is that it's quite an unusual situation when we need to return the entered OldPassword and at the same time we can't find out which password is actually correct. This information is not available and its validation is done in Database.ChangeUserPassword(Text, Text)
https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/database/database-changeuserpassword-method

My original code supported in/out at the same time, which essentially supported the Change Password report scenario. In order to keep this support in the report the old implementation is used instead of the new one.

Let me know what you think about it.

JesperSchulz commented on src/System Application/App/Password/src/PasswordDialogManagement.Codeunit.al:
```diff
@@ -77,6 +77,18 @@ codeunit 9810 "Password Dialog Management"
+    /// </summary>
+    /// <param name="OldPassword">Out parameter, the old password user typed on the dialog.</param>
+    /// <param name="Password">Out parameter, the new password user typed on the dialog.</param>
+    [Obsolete('Replaced by OpenPasswordChangeDialog without out OldPassword param', '24.0')]
```
```suggestion
    [Obsolete('Replaced by OpenPasswordChangeDialog without out OldPassword param', '25.0')]
```

JesperSchulz commented:
@Drakonian, test compilation now seems to fail. Will you investigate?

Drakonian commented:
@JesperSchulz

Yes of course, the problem is that Test Password is a separate app and it cannot access internal implementation from System Application. The strangest thing is that I didn't get this error in testing locally, probably the build system in the project is still not perfect.

I will now update the test that refers to internal codeunit.

grobyns commented:
> @JesperSchulz
> 
> Yes of course, the problem is that Test Password is a separate app and it cannot access internal implementation from System Application. The strangest thing is that I didn't get this error in testing locally, probably the build system in the project is still not perfect.
> 
> I will now update the test that refers to internal codeunit.

FYI: tests should not have direct access to internals. If some internals access is needed, then it should be through a test library app which encapsulates the internal and exposes some public method allowing what you need. If you can avoid using internals completely, that's great, if not, please add methods to a test library :-)

Drakonian commented:
@JesperSchulz 
Ready to check

Drakonian commented:
@JesperSchulz broken builder?

![image](https://github.com/microsoft/BCApps/assets/16802407/aa2a40ba-40d9-4697-97cc-65dc91ae7533)

id looks correct

![image](https://github.com/microsoft/BCApps/assets/16802407/fce41746-0c5b-4859-a097-8a1aeb410f97)

Drakonian commented:
![image](https://github.com/microsoft/BCApps/assets/16802407/70120ac3-c8ea-4ac8-98e2-6c6ab0329c6c)

JesperSchulz commented:
@Drakonian, no, there's still a permission set with ID 135033. Now that you've changed the app.json, that ID is outside of the ID range - hence the compilation error. Add the range from 135033 to 135033, and that issue should be gone.

Drakonian commented:
> @Drakonian, no, there's still a permission set with ID 135033. Now that you've changed the app.json, that ID is outside of the ID range - hence the compilation error. Add the range from 135033 to 135033, and that issue should be gone.

Corrected. 
Do you have any idea why this is compiling locally? 😮

JesperSchulz commented:
> > @Drakonian, no, there's still a permission set with ID 135033. Now that you've changed the app.json, that ID is outside of the ID range - hence the compilation error. Add the range from 135033 to 135033, and that issue should be gone.
> 
> Corrected to old ID range. Do you have any idea why this is compiling locally? 😮

You are probably only compiling the entire System Application in one go, and not the modules individually - is my guess 🙂

Drakonian commented:
> > > @Drakonian, no, there's still a permission set with ID 135033. Now that you've changed the app.json, that ID is outside of the ID range - hence the compilation error. Add the range from 135033 to 135033, and that issue should be gone.
> > 
> > 
> > Corrected to old ID range. Do you have any idea why this is compiling locally? 😮
> 
> You are probably only compiling the entire System Application in one go, and not the modules individually - is my guess 🙂

I feel a little confused about having to run the CI process several times and it finds new errors. It would be cool to run it locally somehow 😞 
I hope you're not wasting too much time on this

new correction ready, should be last 😈

JesperSchulz commented:
No worries. I just click a button 🙂 It's perfectly normal that you need to submit several updates before builds succeed. Of course you can (and should) enable all code analyzers with the correct rules, but things like compiling the modules individually or running CLEAN builds, is too much to ask. We all here in the BC team use CI to verify those things.
